sa/storage-authority.go

```diff
@@ -484,54 +525,102 @@ func (ssa *SQLStorageAuthority) FinalizeAuthorization(authz core.Authorization)
+// will still be mostly empty, but the SA will assign it a request ID.
+func (ssa *SQLStorageAuthority) NewPendingCertificate(cert core.Certificate) (output core.Certificate, err error) {
+	tx, err := ssa.dbMap.Begin()
 	if err != nil {
 		return
 	}
-	digest = core.Fingerprint256(certDER)
-	serial := core.SerialToString(parsedCertificate.SerialNumber)
 
-	cert := &core.Certificate{
-		RegistrationID: regID,
-		Serial:         serial,
-		Digest:         digest,
-		DER:            certDER,
-		Issued:         time.Now(),
-		Expires:        parsedCertificate.NotAfter,
+	// Assign a new request ID that is unique in the database
+	cert.Status = core.StatusPending
+	cert.RequestID = core.NewToken()
+	for existingPendingCert(tx, cert.RequestID) || existingFinalCert(tx, cert.RequestID) {
+		cert.RequestID = core.NewToken()
 	}
-	certStatus := &core.CertificateStatus{
-		SubscriberApproved: false,
-		Status:             core.OCSPStatus("good"),
-		OCSPLastUpdated:    time.Time{},
-		Serial:             serial,
-		RevokedDate:        time.Time{},
-		RevokedReason:      0,
-		LockCol:            0,
+
+	// Insert a stub row in pending
+	pendingCert := pendingCertModel{Certificate: cert}
+	err = tx.Insert(&pendingCert)
+	if err != nil {
+		tx.Rollback()
+		return
+	}
+
+	err = tx.Commit()
+	output = pendingCert.Certificate
+	return
+}
+
+// FinalizeCertificate takes a pending certificate and "finalizes" it, storing
+// a record with a final state ("valid" or "invalid") in the database, and, if
+// "valid", provisioning certificate status information for it.
+//
+// The caller is responsible for populating all the fields in the certificate
+// object before sending it to the SA for storage.
+func (ssa *SQLStorageAuthority) FinalizeCertificate(cert core.Certificate) (err error) {
+	if cert.Status != core.StatusValid && cert.Status != core.StatusInvalid {
+		return fmt.Errorf("Cert must only be finalized to 'valid' or 'invalid' status")
 	}
 
 	tx, err := ssa.dbMap.Begin()
 	if err != nil {
+		return err
+	}
+
+	// Check that a pending certificate exists
+	if !existingPendingCert(tx, cert.RequestID) {
+		err = fmt.Errorf("Cannot finalize a certificate that is not pending")
+		tx.Rollback()
 		return
 	}
 
-	// TODO Verify that the serial number doesn't yet exist
-	err = tx.Insert(cert)
+	// Check that no cert with this serial number exists
+	if existingFinalCertWithSerial(tx, cert.Serial) {
+		// XXX We should probably log this?
+		err = fmt.Errorf("Duplicate serial number detected! [%s]", cert.Serial)
+		tx.Rollback()
+		return
+	}
+
+	// Delete the pending certificate
+	_, err = tx.Delete(&pendingCertModel{cert})
 	if err != nil {
 		tx.Rollback()
 		return
 	}
 
-	err = tx.Insert(certStatus)
+	// Insert the new certificate
+	err = tx.Insert(&cert)
 	if err != nil {
 		tx.Rollback()
 		return
 	}
 
+	// If the certificate is good, add status information
+	if cert.Status == core.StatusValid {
```

Comment:
No.  This method needs to provision an initial record so that `UpdateOCSP()` can update it.